### PR TITLE
Removed AlignedArray 

### PR DESCRIPTION
**Pull request by Anipik**
- Removed Aligned Array from Rff and timeseries. The only place left is FactorAwareFactorization. So I moved it to that namespace so that nobody can use aligned array anymore
- Matmul and MatmulTran modified for any dimension of row and col
- added some more tests for new dimension changes
- corrected the base implementation of MatMulTrans
- added tolerance for some tests for netcoreapp3.0

Fixes https://github.com/dotnet/machinelearning/issues/1018
Related to https://github.com/dotnet/machinelearning/issues/1096

### CODE REVIEW

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ namespace Microsoft.ML.Runtime.Internal.CpuMath
 {
     public static partial class CpuMathUtils
     {
-        public static void MatrixTimesSource(bool transpose, ReadOnlySpan<float> matrix, ReadOnlySpan<float> source, Span<float> destination, int stride)
+        public static void MatrixTimesSource(bool transpose, float[] matrix, ReadOnlySpan<float> source, Span<float> destination, int stride)
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: This should stay a Span. Don't change it back to an array.

File: src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs [hunk 2]
```diff
@@ -295,15 +293,11 @@ public RandomFourierFeaturizingTransformer(IHostEnvironment env, IDataView input
         }
 
         // Round cflt up to a multiple of cfltAlign.
```
Comment (eerhardt) [inline src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs]: This comment is out of date.

File: src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs [hunk 3]
```diff
@@ -571,8 +565,8 @@ private ValueGetter<VBuffer<float>> GetterFromFloatType(IRow input, int iinfo)
                 var getSrc = input.GetGetter<float>(_srcCols[iinfo]);
                 var src = default(float);
 
-                var featuresAligned = new AlignedArray(RoundUp(1, _cfltAlign), CpuMathUtils.GetVectorAlignment());
-                var productAligned = new AlignedArray(RoundUp(_parent._transformInfos[iinfo].NewDim, _cfltAlign), CpuMathUtils.GetVectorAlignment());
+                var featuresAligned = new float[RoundUp(1)];
```
Comment (eerhardt) [inline src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs]: Do you really need to `RoundUp` here anymore?

File: src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs [hunk 4]
```diff
@@ -607,19 +601,25 @@ private ValueGetter<VBuffer<float>> GetterFromFloatType(IRow input, int iinfo)
 
                 if (src.IsDense)
                 {
-                    featuresAligned.CopyFrom(src.Values, 0, src.Length);
+                    Array.Copy(src.Values, featuresAligned, src.Length);
```
Comment (eerhardt) [inline src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs]: This is going to conflict with my change: #1650.  Instead it should be:

`src.GetValues().CopyTo(featuresAligned);`

File: src/Microsoft.ML.TimeSeries/AdaptiveSingularSpectrumSequenceModeler.cs [hunk 5]
```diff
@@ -525,8 +524,7 @@ public override void Save(ModelSaveContext ctx)
             {
                 // REVIEW: this may not be the most efficient way for serializing an aligned matrix.
```
Comment (eerhardt) [inline src/Microsoft.ML.TimeSeries/AdaptiveSingularSpectrumSequenceModeler.cs]: Comment is out of date

File: src/Microsoft.ML.TimeSeries/AdaptiveSingularSpectrumSequenceModeler.cs [hunk 6]
```diff
@@ -1399,8 +1395,8 @@ private void TrainCore(Single[] dataArray, int originalSeriesLength)
                     _x[i - originalSeriesLength + _windowSize] = dataArray[i];
             }
 
-            CpuAligenedMathUtils<CpuAlignedMatrixRow>.MatTimesSrc(_wTrans, _x, _y);
-            CpuAligenedMathUtils<CpuAlignedMatrixRow>.MatTranTimesSrc(_wTrans, _y, _xSmooth);
+            CpuMathUtils.MatrixTimesSource(false, _wTrans, _x, _y, _y.Length);
```
Comment (eerhardt) [inline src/Microsoft.ML.TimeSeries/AdaptiveSingularSpectrumSequenceModeler.cs]: it would be more readable to qualify what the `false` and `true` are:

`CpuMathUtils.MatrixTimesSource(transposed: false, _wTrans, _x, _y, _y.Length);`

File: src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachineInterface.cs [hunk 7]
```diff
@@ -40,53 +24,48 @@ private static bool Compat(AlignedArray a)
...
 
             unsafe
             {
                 fixed (int* pf = &fieldIndices[0])
                 fixed (int* pi = &featureIndices[0])
                 fixed (float* px = &featureValues[0])
                 fixed (float* pw = &linearWeights[0])
-                fixed (float* pv = &latentWeights.Items[0])
-                fixed (float* pq = &latentSum.Items[0])
+                fixed (float* pv = &latentWeights[0])
```
Comment (eerhardt) [inline src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachineInterface.cs]: You should still be asserting above that these arrays are non-empty.

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs (hunk 1 above)
Comment (Anipik) [reply src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: I earlier tried with ROS but some tests are not then finding this function and I am getting Missing Method Exception. I will try to find more info on the reason behind it or source of the error

File: src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs (hunk 3 above)
Comment (Anipik) [reply src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs]: yes this is required for using sse operations, where we will be loading things as a multiple of 4. I can rewriting these Factorizations algos in future to support any type of length but i dont think we should do it in this PR.

Comment (Anipik) [reply src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs]: yes we can change here to 4, sry I understand it as removing the ```RoundUp``` function entirely

File: src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp [hunk 8]
```diff
@@ -48,7 +48,7 @@ EXPORT_API(void) CalculateIntermediateVariablesNative(int fieldCount, int latent
 
         for (int k = 0; k + 4 <= d; k += 4)
         {
-            const __m128 _v = _mm_load_ps(vjf + k);
+            const __m128 _v = _mm_loadu_ps(vjf + k);
```
Comment (wschin) [inline src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp]: This is a major change to Factorization Machine. Do you have any perf result for this new function?

Comment (Anipik) [reply src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp]: I currently don't have any perf results for this, but I will test it.

Comment (wschin) [reply src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp]: I'd suggest using Criteo data set which was used when designing this algorithm.

Comment (Anipik) [reply src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp]: do you have any rsp or test for this dataset ?

Comment (Anipik) [reply src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp]: For this PR i reverted the change to not editing the Factorization Machine

File: test/Microsoft.ML.TestFramework/DataPipe/TestDataPipeBase.cs [hunk 9]
```diff
@@ -1158,12 +1158,12 @@ protected Func<bool> GetColumnComparer(IRow r1, IRow r2, int col, ColumnType typ
...
         }
 
-        private const Double DoubleEps = 1e-9;
+        private const Double DoubleEps = 1e-5;
 
         private static bool EqualWithEpsDouble(Double x, Double y)
         {
             // bitwise comparison is needed because Abs(Inf-Inf) and Abs(NaN-NaN) are not 0s.
-            return FloatUtils.GetBits(x) == FloatUtils.GetBits(y) || Math.Abs(x - y) < DoubleEps;
+            return FloatUtils.GetBits(x) == FloatUtils.GetBits(y) || Math.Abs(x - y) / Math.Abs(x) < DoubleEps;
```
Comment (tannergooding) [inline test/Microsoft.ML.TestFramework/DataPipe/TestDataPipeBase.cs]: Why are you dividing by `abs(x)`?

File: test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/UnitTests.cs [hunk 10]
```diff
@@ -86,29 +71,127 @@ public CpuMathUtilsUnitTests()
...
+        [InlineData(2, 8)]
+        [InlineData(2, 9)]
+        [InlineData(2, 3)]
+        [InlineData(2, 5)]
+        [InlineData(4, 5)]
+        [InlineData(4, 7)]
+        [InlineData(4, 9)]
+        [InlineData(5, 7)]
+        [InlineData(5, 9)]
+
```
Comment (tannergooding) [inline test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/UnitTests.cs]: nit: stray newline

File: src/Native/CpuMathNative/Sse.cpp [hunk 11]
```diff
@@ -60,17 +60,166 @@ const unsigned int TrailingAlignmentMask[16] =
 // Multiply matrix times vector into vector.
 EXPORT_API(void) MatMul(_In_ const float * pmat, _In_ const float * psrc, _Inout_ float * pdst, int crow, int ccol)
 {
+    if (ccol < 4)
+    {
+        for (int i = 0 ; i < crow; i++)
+        {
+            float dotProduct = 0;
+            for (int j = 0; j < ccol; j++)
```
Comment (tannergooding) [inline src/Native/CpuMathNative/Sse.cpp]: nit: unrolling this, since it is only 4 entries, is likely beneficial

File: src/Native/CpuMathNative/Sse.cpp [hunk 12]
```diff
@@ -60,17 +60,166 @@ const unsigned int TrailingAlignmentMask[16] =
...
+            res0 = _mm_hadd_ps(res0, res2);
+
+            _mm_storeu_ps(pDstCurrent, res0);
+
+            pDstCurrent += 4;
+            pMatCurrent += 3 * ccol;
+        }
+    }
+    
     while (pDstCurrent < pDstEnd)
```
Comment (tannergooding) [inline src/Native/CpuMathNative/Sse.cpp]: Why are we only doing one at a time, rather than doing 4 at a time until we can't anymore and then falling back to 1 at a time for the (at most) 3 trailing elements?

File: src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs [hunk 13]
```diff
@@ -572,8 +565,8 @@ private ValueGetter<VBuffer<float>> GetterFromFloatType(IRow input, int iinfo)
                 var getSrc = input.GetGetter<float>(_srcCols[iinfo]);
                 var src = default(float);
 
-                var featuresAligned = new AlignedArray(RoundUp(1, _cfltAlign), CpuMathUtils.GetVectorAlignment());
-                var productAligned = new AlignedArray(RoundUp(_parent._transformInfos[iinfo].NewDim, _cfltAlign), CpuMathUtils.GetVectorAlignment());
+                var featuresAligned = new float[4];
```
Comment (tannergooding) [inline src/Microsoft.ML.Transforms/RandomFourierFeaturizing.cs]: I'm not sure `aligned` is a good name for this anymore (given that the array is not guaranteed to be or stay aligned).

File: test/Microsoft.ML.TestFramework/DataPipe/TestDataPipeBase.cs (hunk 9 above)
Comment (Anipik) [reply test/Microsoft.ML.TestFramework/DataPipe/TestDataPipeBase.cs]: when the numbers are large like in hundreds or thousands,  just checking for an absolute error is causing unnecessary failures. Divinding by ```abs(x)``` gives us the measure of relative error

Comment (tannergooding) [reply test/Microsoft.ML.TestFramework/DataPipe/TestDataPipeBase.cs]: We should probably be updating this to call into the comparison algorithm that was already written, which tries to account for these differences (and also gives a centralized place to update if it needs to be tweaked again in the future).

Comment (Anipik) [reply test/Microsoft.ML.TestFramework/DataPipe/TestDataPipeBase.cs]: are you talking about https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.TestFramework/BaseTestBaseline.cs#L547 comparison algo ?
I dont think that this one also doesn't handle the magnitude part

Comment (tannergooding) [reply test/Microsoft.ML.TestFramework/DataPipe/TestDataPipeBase.cs]: The `CompareNumbersWithTolerance` function factors in both the `integer` and `fractional` part when doing the rounding and comparison.

File: src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp (hunk 8 above)
Comment (wschin) [reply src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp]: Factorization machine still uses aligned memory. Can we move forward without touching it?

**Comment by danmosemsft** [comment]
It would be good to fix the spelling of `CpuAligenedMathUtils.cs`

**Comment by danmosemsft** [comment]
@shauheen this PR will not be completed imminently because @Anipik has been occupied with a customer issue.

**Comment by Anipik** [comment]
> It would be good to fix the spelling of CpuAligenedMathUtils.cs

This file will be removed after this PR

File: src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp (hunk 8 above)
Comment (Anipik) [reply src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp]: @eerhardt @danmosemsft is it okay to keep the just aligned array for factorization machine ?

File: src/Native/CpuMathNative/Sse.cpp (hunk 12 above)
Comment (Anipik) [reply src/Native/CpuMathNative/Sse.cpp]: > rather than doing 4 at a time until we can't anymore and then falling back to 1 at a time for the (at most) 3 trailing elements

we are following this approach when ccol is a multiple of 4 because then if pmat is aligned, will mean that ````pmat  +  k * ccol``` will be aligned and we can use the loadAligned for all four rows which wont be the case if ccol is not a multiple of 4, hence we will have calculate the misalignment for each row independently

File: src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp (hunk 8 above)
Comment (eerhardt) [reply src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp]: If we prove it provides value for FactorizaitonMachine, then yes, let's keep it. But move it into the FactorizationMachine assembly.

Comment (Anipik) [reply src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp]: I have already moved it to FactorizationMachine assembly. The main reason for me to skip is that these algorithms are not as straight forward as cpumathnative functions. Changing them will require a lot of time and a person who understands them completely.

**Comment by Anipik** [comment]
@tannergooding @eerhardt can you please review this one ? i have addressed the feedback

File: test/Microsoft.ML.TimeSeries.Tests/TimeSeriesEstimatorTests.cs [hunk 14]
```diff
@@ -41,7 +41,7 @@ public TimeSeriesEstimatorTests(ITestOutputHelper output) : base(output)
         {
         }
 
-        [ConditionalFact(typeof(BaseTestBaseline), nameof(BaseTestBaseline.LessThanNetCore30OrNotNetCore))] // netcore3.0 output differs from Baseline
+        [ConditionalFact(typeof(BaseTestBaseline), nameof(BaseTestBaseline.LessThanNetCore30OrNotNetCoreAnd64BitProcess))] // 32bit and netcore3.0 output differs from Baseline
```
Comment (eerhardt) [inline test/Microsoft.ML.TimeSeries.Tests/TimeSeriesEstimatorTests.cs]: This change seems unrelated. Why is it necessary?

File: src/Native/CpuMathNative/Sse.cpp [hunk 15]
```diff
@@ -60,17 +60,171 @@ const unsigned int TrailingAlignmentMask[16] =
...
 {
+    if (ccol < 4)
+    {
+        for (int i = 0 ; i < crow; i++)
+        {
+            float dotProduct = 0;
+            switch (ccol)
+            {
+                case 3:
+                    dotProduct += pmat[i * ccol + 2] * psrc[2];
```
Comment (eerhardt) [inline src/Native/CpuMathNative/Sse.cpp]: (minor) It may be beneficial here to note that you are falling through the case statements. I'm not sure...

Comment (Anipik) [reply src/Native/CpuMathNative/Sse.cpp]: yes that is intentional, I can add a comment for it. We are doing a similar thing in scale and add operations

File: test/Microsoft.ML.TimeSeries.Tests/TimeSeriesEstimatorTests.cs (hunk 14 above)
Comment (Anipik) [reply test/Microsoft.ML.TimeSeries.Tests/TimeSeriesEstimatorTests.cs]: Our new algorithm leads to slightly different numbers due to different number of multiplications. (in case of floating points). After the change the x86 baseline numbers are almost same as netcoreapp but different from the original x64

**Comment by Anipik** [comment]
@tannergooding can you please review this one ?

File: src/Native/CpuMathNative/Sse.cpp (hunk 12 above)
Comment (tannergooding) [reply src/Native/CpuMathNative/Sse.cpp]: We could just move the ccol check as part of the `if ((misalignment & 3) != 0)` check.

Operating on 4 columns at a time, aligned or not, will likely perform slightly better and guarantees that we have, at most, 3 trailing elements which makes the trailing code simpler as well (you can just manually unroll it)

Comment (Anipik) [reply src/Native/CpuMathNative/Sse.cpp]: is the unrolling okay ?

```C#
 {
     switch (numRows)
    {
        case 3:
            *(pDstCurrent + 2) =  RowMul(pMatCurrent + 2 * ccol , psrc, pSrcEnd);
        case 2:
            *(pDstCurrent + 1) =  RowMul(pMatCurrent + ccol, psrc, pSrcEnd);
        case 1:
            *pDstCurrent =  RowMul(pMatCurrent, psrc, pSrcEnd);
    }
}
float RowMul(_In_ const float* pMatCurrent, _In_ const float* pSrcCurrent, _In_ const float* pSrcEnd)
{
    __m128 res0 = _mm_setzero_ps();
    while (pSrcCurrent < pSrcEnd)
    {
        __m128 vector = _mm_loadu_ps(pSrcCurrent);

        const float* pMatTemp = pMatCurrent;
        __m128 x01 = _mm_mul_ps(vector, _mm_loadu_ps(pMatTemp));
        res0 = _mm_add_ps(res0, x01);

        pSrcCurrent += 4;
        pMatCurrent += 4;
    }
    
    // Sum all the elements together and return the result
    res0 = _mm_hadd_ps(res0, res0);
    res0 = _mm_hadd_ps(res0, res0);
    return _mm_cvtss_f32(res0);
}

```

Comment (tannergooding) [reply src/Native/CpuMathNative/Sse.cpp]: Wasn't there a code snippet here? I was writing up a response to it...

Comment (Anipik) [reply src/Native/CpuMathNative/Sse.cpp]: sry I edited the snippet. earlier was failing the tests
```C#
    {
                switch (numRows)
                {
                    case 3:
                        *(pDstCurrent + 2) = RowMultiply(pMatCurrent + 2 * ccol, psrc, pSrcEnd, ccol);
                        goto case 2;
                    case 2:
                        *(pDstCurrent + 1) = RowMultiply(pMatCurrent + 1 * ccol, psrc, pSrcEnd, ccol);
                        goto case 1;
                    case 1:
                        *pDstCurrent = RowMultiply(pMatCurrent, psrc, pSrcEnd, ccol);
                        break;
                }
            }
        }

        private  static unsafe float RowMultiply(float* pMatCurrent, float* pSrcCurrent, float* pSrcEnd, int ccol)
        {
            Vector256<float> res0 = Avx.SetZeroVector256<float>();
            int remainder = ccol % 8;
            while (pSrcCurrent + 8 <= pSrcEnd)
            {
                Vector256<float> vector = Avx.LoadVector256(pSrcCurrent);

                float* pMatTemp = pMatCurrent;
                res0 = MultiplyAdd(pMatTemp, vector, res0);

                pSrcCurrent += 8;
                pMatCurrent += 8;
            }

            res0 = VectorSum256(in res0);
            float sum = Sse.ConvertToSingle(Sse.AddScalar(Avx.GetLowerHalf(res0), GetHigh(res0)));

            switch (remainder)
            {
                case 7: sum += *(pSrcCurrent + 6) * *(pMatCurrent + 6); goto case 6;
                case 6: sum += *(pSrcCurrent + 5) * *(pMatCurrent + 5); goto case 5;
                case 5: sum += *(pSrcCurrent + 4) * *(pMatCurrent + 4); goto case 4;
                case 4: sum += *(pSrcCurrent + 3) * *(pMatCurrent + 3); goto case 3;
                case 3: sum += *(pSrcCurrent + 2) * *(pMatCurrent + 2); goto case 2;
                case 2: sum += *(pSrcCurrent + 1) * *(pMatCurrent + 1); goto case 1;
                case 1: sum += *(pSrcCurrent) * *(pMatCurrent); break;
            }

            return sum;
        }
```

**Comment by Anipik** [comment]
@tannergooding I have added the code snippet we talked about in all the implementations
